pytket/extensions/cutensornet/tnstate/ttn.py

```diff
@@ -0,0 +1,845 @@
+from pytket.pauli import QubitPauliString
+
+from pytket.extensions.cutensornet.general import set_logger
+
+from .general import CuTensorNetHandle, Config, TNState, Tensor, _safe_qr
+
+
+class DirTTN(IntEnum):
+    """An enum to refer to relative directions within the TTN."""
+
+    PARENT = -1
+    LEFT = 0
+    RIGHT = 1
+
+
+# An alias for the TTN path from root to a TreeNode
+RootPath = tuple[DirTTN, ...]
+
+
+class TreeNode:
+    """Represents a single tensor in the TTN.
+
+    The shape of the tensor agrees with the convention set in ``DirTTN``, meaning
+    that ``tensor.shape[DirTTN.PARENT]`` corresponds to the dimension of the bond
+    connecting this tree node with its parent. Notice that, since DirTTN.PARENT is
+    -1, this is always the last entry.
+
+    In the case the TreeNode is a leaf, it will contain only one virtual bond
+    (the parent) and as many physical bonds as qubits in the group it represents.
+    These qubits will correspond to bonds from ``0`` to ``len(tensor.shape)-2``.
+    """
+
+    def __init__(self, tensor: Tensor, is_leaf: bool = False):
+        self.tensor = tensor
+        self.is_leaf = is_leaf
+        self.canonical_form: Optional[DirTTN] = None
+
+    def copy(self) -> TreeNode:
+        new_node = TreeNode(
+            self.tensor.copy(),
+            is_leaf=self.is_leaf,
+        )
+        new_node.canonical_form = self.canonical_form
+        return new_node
+
+
+class TTN(TNState):
+    """Represents a state as a Tree Tensor Network.
+
+    Attributes:
+        nodes (dict[RootPath, TreeNode]): A dictionary providing the tree node
+            of the given root path in the TTN.
+        qubit_position (dict[pytket.circuit.Qubit, tuple[RootPath, int]]): A dictionary
+            mapping circuit qubits to their address in the TTN.
+        fidelity (float): A lower bound of the fidelity, obtained by multiplying
+            the fidelities after each contraction. The fidelity of a contraction
+            corresponds to ``|<psi|phi>|^2`` where ``|psi>`` and ``|phi>`` are the
+            states before and after truncation (assuming both are normalised).
+    """
+
+    def __init__(
+        self,
+        libhandle: CuTensorNetHandle,
+        qubit_partition: dict[int, list[Qubit]],
+        config: Config,
+    ):
+        """Initialise a TTN on the computational state ``|0>``.
+
+        Note:
+            A ``libhandle`` should be created via a ``with CuTensorNet() as libhandle:``
+            statement. The device where the TTN is stored will match the one specified
+            by the library handle.
+
+            The current implementation requires the keys of ``qubit_partition`` to be
+            integers from ``0`` to ``2^l - 1`` for some ``l``.
+
+        Args:
+            libhandle: The cuTensorNet library handle that will be used to carry out
+                tensor operations on the TTN.
+            qubit_partition: A partition of the qubits in the circuit into disjoint

```

Comment:
Perhaps should explain what the integer key means...

Reply:
Done (appearing in next commit)